[SPARK-44394][CONNECT][WEBUI] Add a Spark UI page for Spark Connect

jasonli-db opened the pull request:
## What changes were proposed in this pull request?
Add a new Spark UI page to display session and execution information for Spark Connect. This builds of the work in SPARK-43923 (https://github.com/apache/spark/pull/41443) that adds the relevant SparkListenerEvents and mirrors the ThriftServerPage in the Spark UI for JDBC/ODBC.

<img width="1709" alt="Screenshot 2023-07-27 at 11 29 22 PM" src="https://github.com/apache/spark/assets/65624911/934b7c69-3b44-460b-8fbb-36a9eb3f0798">

<img width="1716" alt="Screenshot 2023-07-27 at 11 29 15 PM" src="https://github.com/apache/spark/assets/65624911/33dbe6ab-44bf-49a5-ad4c-5ba4a476a1f0">


### Why are the changes needed?
This gives users a way to access session and execution information for Spark Connect via the UI and provides the frontend interface for the related SparkListenerEvents.

### Does this PR introduce _any_ user-facing change?
Yes, it will add a new tab/page in the Spark UI


### How was this patch tested?
Unit tests

gengliangwang commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala:
```diff
@@ -115,4 +115,17 @@ object Connect {
+      .doc("The number of statements kept in the Spark Connect UI history.")
+      .version("3.5.0")
+      .intConf
+      .createWithDefault(200)
```
Why the default value is 200? For SQL query retention, the default value is 1000

jasonli-db replied:
This is what the Thrift Server has for defaults: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L1414. Can be changed though. Is 1000 preferred?

jasonli-db commented:
@juliuszsompolski @grundprinzip Can you help review this PR? This builds off of https://github.com/apache/spark/pull/41443, so the commit to look at is https://github.com/apache/spark/pull/41964/commits/fcf0ca69ea30a71e20ec15673d62bbfe91f11a36. 

Also cc @rednaxelafx @gengliangwang

gengliangwang commented:
Since the Spark connect is more about SQL, shall we show the SQL execution links on the UI?

juliuszsompolski commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala:
```diff
@@ -259,13 +281,15 @@ object SparkConnectService {
   // Starts the service
-  def start(): Unit = {
-    startGRPCService()
+  def start(sparkContextOpt: Option[SparkContext] = None): Unit = {
```
make it `sparkContext: SparkContext` , not an option, and in SparkConnectServer pass it explcitly from the session there. better that than having to `.getOrElse(SparkSession.active.sparkContext)` without controlling if there's an active session here

juliuszsompolski commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala:
```diff
@@ -49,6 +49,7 @@ object SparkConnectServer extends Logging {
       SparkConnectService.server.awaitTermination()
     } finally {
       session.stop()
+      SparkConnectService.uiTab.foreach(_.detach())
```
wouldn't SparkConnectService.stop() have been called already somewhere and already detached it?

jasonli-db replied:
Does it get called on this path? I only saw `SparkConnectService.server.awaitTermination()` on line 49. When I searched for SparkConnectService.stop(), I only saw two occurrences in [SimpleSparkConnectService](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/SimpleSparkConnectService.scala#L55) and [SparkConnectPlugin](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/SparkConnectPlugin.scala)

juliuszsompolski replied:
yeah... it seems that in this path the Server is killed by a signal (via `sbin/stop-connect-server.sh`) and then `SparkConnectService.stop()` is never called.
It would be nice to SparkConnectService.stop() here, but then it will try to shutdown the server again... 
In any case, it's unrelated to this PR, so let's leave it like it is now.

juliuszsompolski commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala:
```diff
@@ -259,13 +281,15 @@ object SparkConnectService {
+
     server = sb.build
     server.start()
+    createListenerAndUI(sc)
```
nit: maybe move `createListenerAndUI` to `start`, just like `detach` is in `stop`. Then don't need to pass sc to `startGRPCService`

jasonli-db replied:
Updated.

juliuszsompolski commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala:
```diff
@@ -179,6 +182,9 @@ object SparkConnectService {
 
   private[connect] var server: Server = _
 
+  var uiTab: Option[SparkConnectServerTab] = None
+  var listener: SparkConnectServerListener = _
```
could both be private, or `private[connect]`?

jasonli-db replied:
Updated to `private[connect]`.

juliuszsompolski reviewed (commented):
reviewed connect service parts, didn't review `ui` package.

gengliangwang commented:
> Since the Spark connect is more about SQL, shall we show the SQL execution links on the UI?

@jasonli-db What is your opinion of this one?

jasonli-db commented:
> > Since the Spark connect is more about SQL, shall we show the SQL execution links on the UI?
> 
> @jasonli-db What is your opinion of this one?
 I agree that adding the SQL execution links would be ideal and useful for the user. @juliuszsompolski Can you give your input on this and confirm whether or not this is currently possible. Do we have a link between SQL execution IDs and Connect requests?

juliuszsompolski commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala:
```diff
@@ -0,0 +1,361 @@
+      .toSeq
+      .filter(!_.isEmpty)
+      .sorted
+    val execList = executionList.values.filter(exec => jobTags.contains(exec.jobTag)).toSeq
```
@jasonli-db could you try to use https://github.com/juliuszsompolski/apache-spark/commit/639d515d629d78825489510eb98b3648e4122b04
to make it
```
val jobTag = jobTags.findFirst({ case ExecuteJobTag(jobTag) => jobexeclistTag }
val exec = executionList.get(jobTag)
```

juliuszsompolski commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala:
```diff
@@ -0,0 +1,369 @@
+  }
+
+  override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
+    val executionIdOpt: Option[String] = Option(jobStart.properties)
+      .flatMap { p => Option(p.getProperty(SQLExecution.EXECUTION_ID_KEY)) }
```
Discussed offline with @jasonli-db and @jdesjean that for executions that don't start a Spark Job (e.g. SELECT 1), they will not be recorded. For that reason, as a followup, it would be best if jobTags could also be added to the SparkListenerSQLExecutionStart event.

juliuszsompolski reviewed (approved):
Reviewed and LGTM on the connect part; didn't review details of the UI and state store part.

jasonli-db committed:
Add a Spark UI page for Spark Connect

gengliangwang commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala:
```diff
@@ -0,0 +1,130 @@
+
+  @JsonIgnore @KVIndex("isExecutionActive")
+  def isExecutionActive: Boolean = {
+    !(state == ExecutionState.FAILED ||
```
Shall we use allow list to make it more robust?
E.g. check the state is among STARTED, COMPILED, READY

jasonli-db replied:
Updated.

gengliangwang commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerHistoryServerPlugin.scala:
```diff
@@ -0,0 +1,41 @@
+    }
+  }
+
+  override def displayOrder: Int = 1
```
increase the display order to avoid conflict with StreamingQueryHistoryServerPlugin?

rednaxelafx replied:
It'd be good to have this, but because of how other SHS plugins are also broken in the same way, it may be clearer to have a separate PR that fixes the displayOrder of all of the plugins in one go so that we can see the context of why the values were chosen in one single PR.

e.g. if we set this plugin's display order to `4`, it wouldn't be very obvious just from this PR alone why it should be a `4` while a lot of other plugins have a display order of `1`

jasonli-db replied:
Agreed we should update them all at once. Should we do it in a separate PR or just update all of them in this one?

rednaxelafx replied:
I prefer doing the wholesale update in a separate PR, but please check with the Apache Spark committers in this thread to confirm 🙏

jasonli-db replied:
Hm, I checked and since this is only incorrect in one other place (StreamingQueryHistoryServerPlugin.scala), I just updated it in this PR

jasonli-db committed:
Address comments and fix tests

gengliangwang commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerPage.scala:
```diff
@@ -0,0 +1,529 @@
+    val ordering: Ordering[SqlStatsTableRow] = sortColumn match {
+      case "User" => Ordering.by(_.executionInfo.userId)
+      case "Job Tag" => Ordering.by(_.executionInfo.jobTag)
+      case "Job ID" => Ordering by (_.jobId.headOption)
```
<img width="354" alt="image" src="https://github.com/apache/spark/assets/1097932/62c47d17-56b2-4187-ab59-f96a5af1b240">
Is the ordering working?

gengliangwang commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerTab.scala:
```diff
@@ -0,0 +1,56 @@
+private[connect] class SparkConnectServerTab(
+    val store: SparkConnectServerAppStatusStore,
+    sparkUI: SparkUI)
+    extends SparkUITab(sparkUI, "connect")
```
nit: two spaces indent

jasonli-db replied:
This seems to be set by the linter (it reverted back to this form after I ran the lint command).

gengliangwang commented:
@jasonli-db Please try more tests. I did a quick test and find the SQL IDs need deduplication.
<img width="2526" alt="image" src="https://github.com/apache/spark/assets/1097932/975760a1-c60c-451c-b74d-2a36e7b4a7d4">

gengliangwang commented:
Do we expect users to set the job tags? The generated tags are long and meaningless
<img width="1996" alt="image" src="https://github.com/apache/spark/assets/1097932/88d105a3-4bb8-4241-a9b4-987f78520b0a">

gengliangwang commented:
In the screenshot above, the close time and finish time column seems duplicated. And the execution time and duration also seems quite similar. Why do we have so many columns about the timings?

gengliangwang commented:
Nit: we can have follow-up to improve the page when a session doesn't exist(e.g. evicted from KV store):
<img width="750" alt="image" src="https://github.com/apache/spark/assets/1097932/1171802f-942e-424e-8d2c-154da05ece1e">

For example, id not existed is handled in the SQL page:
<img width="1041" alt="image" src="https://github.com/apache/spark/assets/1097932/8adea10a-e653-40b7-8190-a175301ff41f">

jasonli-db commented:
> Do we expect users to set the job tags? The generated tags are long and meaningless <img alt="image" width="1996" src="https://user-images.githubusercontent.com/1097932/256669420-88d105a3-4bb8-4241-a9b4-987f78520b0a.png">

@juliuszsompolski @jdesjean Is there any use in including the job tags? Actually now that I think about it, since it's constructed from the SessionID, userID, and operationID, I think we can instead just add a column for the operationID and remove the jobTag column. What do you think?

juliuszsompolski commented:
OperationId could indeed maybe be more useful instead. If a user uses connect api like SparkSession.interruptOperation, they pass just operationId.
The only time a user would want to use jobTag is if they wanted to directly call SparkContext.interruptJobsWithTag, which is a rather internal use.
We could consider displaying SparkListenerConnectOperationStarted.sparkSessionTags, because these are uses settable (via connect SparkSession.addTag / removeTag / clearTags), and they are used by users to call connect SparkSession.interruptTag.

juliuszsompolski commented:
> In the screenshot above, the close time and finish time column seems duplicated. And the execution time and duration also seems quite similar. Why do we have so many columns about the timings?

@gengliangwang Finish time is finish execution, close time is when all the results were transferred; execution time is the time query was executing, while duration is end to end. It is meaningful for queries with large results, or e.g. complex queries that spend a lot of time in optimizer. It's the same with Thriftserver UI.

jasonli-db commented:
> @jasonli-db Please try more tests. I did a quick test and find the SQL IDs need deduplication. <img alt="image" width="2000" src="https://user-images.githubusercontent.com/1097932/256668978-975760a1-c60c-451c-b74d-2a36e7b4a7d4.png">

Thanks for catching this. It's been fixed

jasonli-db commented:
> Nit: we can have follow-up to improve the page when a session doesn't exist(e.g. evicted from KV store): <img alt="image" width="750" src="https://user-images.githubusercontent.com/1097932/256671033-1171802f-942e-424e-8d2c-154da05ece1e.png">
> 
> For example, id not existed is handled in the SQL page: <img alt="image" width="1041" src="https://user-images.githubusercontent.com/1097932/256671131-8adea10a-e653-40b7-8190-a175301ff41f.png">

Addressed to match the behavior for the SQL page.

jasonli-db commented:
> OperationId could indeed maybe be more useful instead. If a user uses connect api like SparkSession.interruptOperation, they pass just operationId. The only time a user would want to use jobTag is if they wanted to directly call SparkContext.interruptJobsWithTag, which is a rather internal use. We could consider displaying SparkListenerConnectOperationStarted.sparkSessionTags, because these are uses settable (via connect SparkSession.addTag / removeTag / clearTags), and they are used by users to call connect SparkSession.interruptTag.

@juliuszsompolski @gengliangwang I've updated the screenshots. I added the operationID and sparkSessionTags. For now, I kept the jobTag since it seems it could still be used if only internally. I'll remove it if/when we come to a final consensus to do so. Please take a look!

gengliangwang commented:
@jasonli-db I have to say to the current table is quite wide. 
<img width="1720" alt="image" src="https://github.com/apache/spark/assets/1097932/db421554-d98f-497f-a913-56b1a511e17c">

Can we at least put the operationID column behind the column "State" since it is not commonly useful?


A good way to resolve this is to have checkboxes and make operationID/sparkSessionTags/jobTag as  additional metrics:
<img width="383" alt="image" src="https://github.com/apache/spark/assets/1097932/0195d334-70d5-4fee-88ed-1951242a3692">
This can be done in a followup

juliuszsompolski commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala:
```diff
@@ -0,0 +1,131 @@
+  }
+}
+
+private[connect] object ExecutionState extends Enumeration {
+  val STARTED, COMPILED, READY, CANCELED, FAILED, FINISHED, CLOSED = Value
+  type ExecutionState = Value
+}
```
For potential followup:
I find it a bit weird, that it doesn't have a RUNNING execution state, but it's actually the same as in Thriftserver:
https://github.com/apache/spark/blob/master/sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala#L119C58-L119C58

Spark Connect records that it's "ready for execution" at the end of planner (or before eager commands), and we record "finished" when the result stage completes... Should we have another event "RUNNING" that e.g. for SparkConnectPlanExecution would be right before `spark.sparkContext.submitJob`, and for Commands would need to be determined on a per command basis. WDYT @jdesjean ; that would be different from Thriftserver, but it seems to me like that "RUNNING" state is missing from Thriftserver... cc @wangyum @yaooqinn  @bogdanghit WDYT from the Thriftserver perspective?

bogdanghit replied:
I think a `RUNNING` state should cover the period between it enters `STARTED` state until it enters one of the terminal states. Using this definition, in ThriftServer we can infer the execution is in running / active state if it's not `FAILED, CANCELED, TIMEDOUT or CLOSED.`

Nit: It would be good to keep these states consistent. The SparkConnect execution state has a `READY` state which is missing in ThriftServer.

juliuszsompolski replied:
@bogdanghit see https://github.com/apache/spark/pull/41748 from @jdesjean . Jean did some work to be able to figure out the "READY" state to come before the start of eager commands. In Thriftserver there's a bit of ugly issue that eager commands would execute even before the COMPILED state, because they are eagerly executed straight in analysis.
Thriftserver could use https://github.com/apache/spark/pull/41748 to also display information about eager execution better.

jasonli-db replied:
I have the UI display the state as RUNNING is isActiveExecution is true, which covers STARTED, COMPILED, READY: https://github.com/apache/spark/pull/41964/commits/8abdecdc7cedb6ff5a13090dd4bcae2ecf50d0d0#diff-ef2b5b937c2c5e83d6ca9804c9d402d263bf802550cc8daeead089498c123aa0L322.

jasonli-db committed:
Show RUNNING as the state in the UI if the execution is active

juliuszsompolski commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerPage.scala:
```diff
@@ -0,0 +1,541 @@
+        </span>
+      </td>
+      <td>
+        {if (info.isExecutionActive) "RUNNING" else info.state}
```
LGTM, I think it makes more sense, thanks!

gengliangwang commented:
@jasonli-db Thanks for the work. I will merge the PR once the tests are passed.

jasonli-db committed:
Address issue where handler can be attached more than once

gengliangwang commented:
Thanks, merging to master/3.5

gengliangwang commented:
@jasonli-db This PR has conflict with the branch-3.5
Please cherry-pick it before the RC1 on Aug 1

juliuszsompolski commented:
Thanks @gengliangwang . Since the conflict was trivial (imports), I resolved and raised https://github.com/apache/spark/pull/42224 for branch-3.5.
